### PR TITLE
122: polishing contributing doc

### LINKED ISSUE #122: Polish minor edges in contributing guide
To fix:

image
-> make it an R chunk

image
-> use code formatting

image
-> both links have the dot in front of the section, this must be removed otherwise the link does not work correctly to go to the section on the page

Plus: I would add a convention on how to write comments.
- Should follow sentence style.
- Should be as minimal as possible - generally the code and variable names should be clear enough to not need comments. Only use them when needed

### PR DESCRIPTION
**Pull request by danielinteractive**
closes #122

### CODE REVIEW

File: .github/CONTRIBUTING.Rmd [hunk 1]
```diff
@@ -0,0 +1,185 @@
+---
+output: github_document
+---
+<!-- markdownlint-disable-file -->
+<!-- CONTRIBUTING.md needs to be generated from CONTRIBUTING.Rmd. Please edit that file -->
+
+# Contributing
+
+Thank you for your interest in contributing to this repo!
+Your contribution is highly valued, and please go through this document to help you contribute.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:10]:
```suggestion
Your contribution is highly valued. Please go through this document for guidance on how to contribute.
```

File: .github/CONTRIBUTING.Rmd [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+## Programming conventions
+
+Please follow the programming conventions to ensure a consistent program style across the package.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:14]:
```suggestion
Please follow the programming conventions to ensure a consistent programming style across the package.
```

File: .github/CONTRIBUTING.Rmd [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+Generally we follow the [tidyverse style guide](https://style.tidyverse.org/).
+Some specific conventions that deviate from this are explained here.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:17]:
```suggestion
Some specific conventions that deviate from this are explained below.
```

File: .github/CONTRIBUTING.Rmd [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+### Functions
+
+1. Function names should be explicit and clear, separated by underscore (`snake_case`).
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:21]:
```suggestion
1. Function names should be explicit and clear. Words should be separated by underscore (`snake_case`).
```

File: .github/CONTRIBUTING.Rmd [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  explain the "Why" and not the "What".
+
+## Github conventions
+
+When using github to collaborate, the following conventions are needed:
+
+1. Github issues is for issues, feature requests, bugs. Before creating a issue, please make sure this issue has not been reported yet.
+   - If you are going to work on this issue, please assign yourself.
+2. Please create a branch in the `mmrm` repository, instead of creating forks, unless you are not yet a team member.
+   - The name of a branch should be like `<issue_id>_<short_discription>`. Use an underscore to separate the description.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:40]:
```suggestion
   - Branches should be associated with a GitHub issue and linked to an issue id. 
   - The name of a branch should be of the form: `<issue_id>_<short_discription>`.
```
just to introduce that branches should be linked to issues.

File: .github/CONTRIBUTING.Rmd [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+The development this `mmrm` package is based on the latest R version and C++ compilers.
+The package dependencies are the most recent versions from CRAN.
+We recommend that your working environment is set up in the same way.
+Additionally, there are some tools we recommend you to install:
+
+1. [`RTools`](https://cran.r-project.org/bin/windows/Rtools/rtools42/rtools.html) if you work on a Windows operating system. Alternatively you can use [`docker`](https://www.docker.com/) to separate the operating system and the development system.
+2. [`GitKraken`](https://www.gitkraken.com/) is a very useful user interface for git including visualization of git commit graphs, file history, etc.
+3. [`lintr`](https://openpharma.github.io/mmrm/main/articles/package_structure.html#lintr) will allow you to perform static code analysis.
+4. [`pre-commit`](https://openpharma.github.io/mmrm/main/articles/package_structure.html#pre-commit-config.yaml) is a Python module that allow you to identify issues before you commit locally.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:64]:
```suggestion
4. [`pre-commit`](https://openpharma.github.io/mmrm/main/articles/package_structure.html#pre-commit-config.yaml) is a Python module that allows you to identify issues before you commit locally.
```

File: .github/CONTRIBUTING.Rmd [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+| `SPx`              | `SP` (story points) indicate complexity, and the larger the subsequent number, the more time consuming the issue is expected to be |
+| `priority`         | Issues with this label should be completed with higher priority                                                                    |
+| `good first issue` | Good choices for new team members                                                                                                  |
+| `blocked`          | Blocked by other issues                                                                                                            |
+| `bug`              | Something isn't working                                                                                                            |
+| `devops`           | Development and Operation                                                                                                          |
+| `discussion`       | Disccusion needed                                                                                                                  |
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:78]:
```suggestion
| `discussion`       | Discussion needed                                                                                                                  |
```

File: .github/CONTRIBUTING.Rmd [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+Does the test fit in the scope of existing tests?
+If so, please modify the existing test files under `tests/testthat/` folder or `src/` folder, depending on whether the code to be tested is R or C++.
+Otherwise please create a new test file, with a name prefix of "test-".
+
+In each test case, use the following structure:
+
+```{r eval=FALSE}
+test_that("<function_name> do something as explained", {
+  # test body
+  expect_identical(1, 1)
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:100]: funnily enough, I've provided boilerplate like this to contributors and eventually found `expect_identical(1,1)` in their eventual first PR.

File: .github/CONTRIBUTING.Rmd [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+In the test body part, conduct the tests, e.g. use `expect_identical` to check consistency, `expect_error` to catch error messages, etc.
+The test body should not follow the same implementation logic as the package did, otherwise we may miss mistakes in implementation.
+
+### Add integration tests
+
+Integration tests compare the results of SAS and R and assures the quality of our code.
+To add an integration test, you need to do the following:
+
+1. Use SAS to run an appropriate mmrm models with `proc mixed`, using `fev_data`.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:114]:
```suggestion
1. Use SAS to run an appropriate mmrm model with `proc mixed`, using `fev_data`.
```

File: .github/CONTRIBUTING.Rmd [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+2. Save the results in `txt` format in the `design/SAS/` folder.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:115]:
```suggestion
2. Save the results in `.txt` format in the `design/SAS/` folder.
```

File: .github/CONTRIBUTING.Rmd [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+In `mmrm` we are not including any latent variables and so the Laplace approximation aspect of `TMB` is not used. We only use the automatic differentiation part of `TMB`. For detailed documentation of `TMB`, visit the [TMB reference](https://kaskr.github.io/adcomp/Introduction.html).
+
+One important feature of `TMB` are the R style matrix/array calculations. This is important because we mainly use this part to conduct our calculations. See [matrix_arrays.cpp](https://kaskr.github.io/adcomp/matrix_arrays_8cpp-example.html) for examples.
+
+### Add a new covariance structure
+
+To add a new covariance structure, you need to do the following:
+
+1. Understand the covariance structure and add appropriate documentation in [covariance structure](https://openpharma.github.io/mmrm/main/articles/covariance.html) and you can create a draft pull request to invite discussion from other team members.
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:149]:
```suggestion
1. Understand the covariance structure and add appropriate documentation in [covariance structure](https://openpharma.github.io/mmrm/main/articles/covariance.html). Create a draft pull request to invite discussion from other team members.
```

**Review by BrianLang — COMMENTED** [review]
Have some more phrasing suggestions for us. I think it looks pretty good. Might want to consider if GitHub actions grammar is really that useful... most of the feedback I see is quite off-base.

File: .github/CONTRIBUTING.Rmd (hunk 8 above)
Comment (danielinteractive) [inline .github/CONTRIBUTING.Rmd:100]:
```suggestion
  result <- <function_name>(input)
  expected <- <hardcoded result>
  expect_identical(result, expected)
```

Comment (danielinteractive) [reply .github/CONTRIBUTING.Rmd:100]: thanks, I made it even more obvious now :-)

File: .github/CONTRIBUTING.Rmd [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+### Add new unit tests
+
+To add a new unit test, you need to first identify the test scope.
+Does the test fit in the scope of existing tests?
+If so, please modify the existing test files under `tests/testthat/` folder or `src/` folder, depending on whether the code to be tested is R or C++.
+Otherwise please create a new test file, with a name prefix of "test-".
+
+In each test case, use the following structure:
+
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:97]:
```suggestion

# nolint start 
```

File: .github/CONTRIBUTING.Rmd [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+```{r eval=FALSE}
+test_that("<function_name> do something as explained", {
+  result <- <function_name>(input)
+  expected <- <hardcoded result>
+  expect_identical(result, expected)
+})
+```
+
```
Comment (BrianLang) [inline .github/CONTRIBUTING.Rmd:105]:
```suggestion

# nolint end
```

**Review by BrianLang — APPROVED** [review]
Looks good now!

**Pull request merged**